core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -627,12 +627,22 @@ protected Object visitInPredicate(InPredicate node, Object context)
 
             boolean hasUnresolvedValue = value instanceof Expression;
             boolean hasNullValue = false;
+            boolean hasOnlyLiterals = true;
             boolean found = false;
             List<Object> values = new ArrayList<>(valueList.getValues().size());
             List<Type> types = new ArrayList<>(valueList.getValues().size());
 
             ResolvedFunction equalsOperator = metadata.resolveOperator(session, OperatorType.EQUAL, types(node.getValue(), valueList));
             for (Expression expression : valueList.getValues()) {
+                hasOnlyLiterals &= expression instanceof Literal;
```

Comment:
typically `isEffectivelyLiteral` rather than `instanceof Literal`

same below

Reply:
Document in the code why you chose not to use `isEffectivelyLiteral`.
Note that currently there are pretty much _none_ occurrences of  `instanceof Literal`, all have been replaced with `isEffectivelyLiteral`, so doing `instanceof Literal` requires a commentary.

Reply:
> Document in the code why you chose not to use isEffectivelyLiteral.

It is documented:
```
                    // skip interpreting of literal IN term since it cannot be compared
                    // with unresolved "value" and it cannot be simplified further
```
Expressions for which `isEffectivelyLiteral` returns `true` can be simplified further

Reply:
> Expressions for which `isEffectivelyLiteral` returns `true` can be simplified further

That's true. Can you make it explicit within this code comment?